### PR TITLE
Implement NULLCHK Evaluator for AArch64

### PR DESCRIPTION
**Pull request by aarongraham9**
Implement NULLCHK Evaluator for AArch64

Signed-off-by: Aaron Graham <aaron.graham@unb.ca>

### CODE REVIEW

**Comment by aarongraham9** [comment]
Do we want to worry about the `PossibleResolve` at this point @0xdaryl?

**Comment by 0xdaryl** [comment]
This is a far more complicated solution (450 LOC!) than is required at this point (in my opinion).  Plus, this is an exact cut-and-paste of the Z version and will not build nor work on AArch64.

Only explicit NULLCHKs need to be supported at this point.  This essentially means that you do a NULLness check on the base object for the tree in question and if it is NULL then branch to some sort of helper snippet to throw the AIOB exception.  It looks like Power has a fairly readable implementation so I suggest you start with that for how you need to implement this.  Just follow the path of where an explicit check is required.  The x86 implementation isn't too bad to follow either.

Other than the actual checking and snippet generation parts, you will need to handle evaluating and decrementing the various nodes in the tree.  See how the example implementations do it.

Don't worry about the resolution code at this point (i.e., `TR::resolveAndNULLCHK` nodes).  I believe most platforms have this incorrect already (see eclipse/openj9#6347) and it will really only be a problem in situations that are specifically testing for that behaviour.  We can fix that functionality up later.

A simple solution that only does explicit checks and branches to a snippet to call some VM helper to throw an exception can be implemented in OMR.  However, if resolution enters the picture I think the solution should be done in OpenJ9 (but as I mentioned, I don't think you need to worry about the resolution case at this point).

**Comment by aarongraham9** [comment]
Yes it won't build/work, this version wasn't ready for review yet. I'll have an updated version for tomorrow.

I agree this version is overkill. Ok, I'll cut it down to the explicit case only for now. I saw eclipse/openj9#6347; sounds good. I'll incorporate your feedback and get this updated tomorrow.

Thanks!

**Comment by aarongraham9** [comment]
Local `AArch64` build and Tril tests passed. This is ready for review @0xdaryl.

File: compiler/aarch64/codegen/ControlFlowEvaluator.cpp [hunk 1]
```diff
@@ -480,12 +480,65 @@ OMR::ARM64::TreeEvaluator::tableEvaluator(TR::Node *node, TR::CodeGenerator *cg)
...
+   // in which case some kind of explicit test or indirect load must be generated
+   // to force the null check at this point.
+
+   TR::Node *firstChild = node->getFirstChild();
+   TR::ILOpCode &opCode = firstChild->getOpCode();
+   TR::Node *reference = NULL;
+   TR::Compilation *comp = cg->comp();
+
+   reference = node->getNullCheckReference();
+
```
Comment (0xdaryl) [inline compiler/aarch64/codegen/ControlFlowEvaluator.cpp]: You are missing the part that actually does the NULLness check and branches to a snippet for handling.  For an explicit check, right about here you will need code that looks something like this:
```
   TR::LabelSymbol *snippetLabel = generateLabelSymbol(cg);
   cg->addSnippet(new (cg->trHeapMemory()) TR::ARM64HelperCallSnippet(cg, node, snippetLabel, node->getSymbolReference()));

   TR::Register *referenceReg = cg->evaluate(reference);
   TR::InstOpCode::Mnemonic compareOp = comp->useCompressedPointers() ? TR::InstOpCode::cbzw : TR::InstOpCode::cbzx;
   TR::Instruction *cbzInstruction = generateCompareBranchInstruction(cg, compareOp, node, divisorReg, snippetLabel, NULL);

   // consider all registers live
   cbzInstruction->setNeedsGCMap(0xffffffff);
```

This code will have to reconcile with the logic that you implemented below (especially the node evaluation).

Also, have a look at what Power and X does for compressed refs when the node is an `l2a` and how you have to reach for the actual reference.  You'll probably need something like that as well.

**Comment by knn-k** [comment]
#4128 for `resolveAndNULLCHKEvaluator()` calls the function `evaluateNULLCHKWithPossibleResolve()` introduced by this PR.

**Comment by 0xdaryl** [comment]
Any progress on this?

**Comment by 0xdaryl** [comment]
Any progress @aarongraham9 ?  Thanks.

**Comment by aarongraham9** [comment]
Not as of yet. Just back from vacation and I'm catching back up. I'll get to it as soon as I'm able.

File: compiler/aarch64/codegen/OMRTreeEvaluator.hpp [hunk 2]
```diff
@@ -49,6 +49,14 @@ namespace TR { class CodeGenerator; }
  */
 TR::Register *genericReturnEvaluator(TR::Node *node, TR::RealRegister::RegNum rnum, TR_RegisterKinds rk, TR_ReturnInfo i, TR::CodeGenerator *cg);
 
+/**
+ * @brief Helper function for Null/Resolve CHK Evaluators
+ * @param[in] node : node
+ * @param[in] needsResolve : if node needs resoloution
+ * @param[in] cg : CodeGenerator
+ */
+TR::Register *evaluateNULLCHKWithPossibleResolve(TR::Node *node, bool needsResolve, TR::CodeGenerator *cg);
```
Comment (knn-k) [inline compiler/aarch64/codegen/OMRTreeEvaluator.hpp]: I don't think you need to declare this function in OMRTreeEvaluator.hpp.
Are you planning to call `evaluateNULLCHKWithPossibleResolve()` from files other than ControlFlowEvaluator.cpp?

**Comment by aarongraham9** [comment]
This PR (#4091) has been superseded by #4187. Closing.

**Pull request closed without merging**